cmd/kubelet/app/server.go

```diff
@@ -285,6 +300,14 @@ HTTP server: The kubelet can also listen for HTTP and respond to a simple API
 	return cmd
 }
 
+// mergeKubeletConfiguration merges specific fields from KubeletInstanceConfiguration to KubeletConfiguration
+func mergeKubeletConfiguration(kc *kubeletconfiginternal.KubeletConfiguration, kic *kubeletconfiginternal.KubeletInstanceConfiguration) *kubeletconfiginternal.KubeletConfiguration {
+	if kc.RegisterNode != kic.RegisterNode {
```

Comment:
What happens when there are two fields in the instance configuration, and one isn't explicitly set in the instance config, but is explicitly set to false in the shared config?

Example types:
```
type Instance struct {
  Foo bool // defaults true
  Bar bool // defaults false
}

type Shared struct {
  Foo bool // defaults true
}
```

Set as follows:
```
kind: Instance
bar: true
---
kind: Shared
foo: false
```

Read in and default to:
```
i := &Instance{
  Foo: true, // but the user didn't specify an override for `foo` in the instance config file
  Bar: false,
}

s := &Shared{
  Foo: false, 
}
```

Merge result:
```
i := &Instance{
  Foo: true, 
  Bar: false,
}

s := &Shared{
  Foo: true, // not the user's intent
}
```

We need a way to track whether the `foo` was actually set by the user when we do the merge.

Reply:
With the refactor, we have:

Example types:

```
type Instance struct {
  Foo bool // defaults nil
  Bar bool // defaults false
}

type Shared struct {
  Foo bool // defaults true
}
```

Set as follows:

```
kind: Instance
bar: true
---
kind: Shared
foo: false
```

Read in and default to:

```
i := &Instance{
  Foo: false, // the value is copied from the shared instance
  Bar: true,
}

s := &Shared{
  Foo: false, 
}
```

Merge result:

```
i := &Instance{
  Foo: false, 
  Bar: true,
}

s := &Shared{
  Foo: false,
  Bar: true,
}
```

Reply:
With this new flow:

```
type Instance struct {
  Foo bool // defaults nil
  Bar bool // defaults nil
}

type Shared struct {
  Foo bool // defaults true
}
```

Set as follows:

```
kind: Instance
bar: true
---
kind: Shared
foo: false
```

Read in and default to:

```
i := &Instance{
  Foo: false, // this is copied from the final shared value
  Bar: true, // from the YAML
}

s := &Shared{
  Foo: false, 
}
```

Merge result:

```
i := &Instance{
  Foo: false, 
  Bar: true,
}

s := &Shared{
  Foo: false,
}
```

Reply:
Good idea. Benefit is we don't need to rely on fields being pointers.

To summarize my understanding of how we'd implement this:

1. Decode `shared.yaml` to `SharedExternal`.
1. Construct `InstanceExternal` such that any fields that exist in both types are set to the values from `SharedExternal`.
1. Decode `instance.yaml` over `InstanceExternal`. Values set in `shared.yaml` but not `instance.yaml` will match intent from `shared.yaml` due to the previous copy from `SharedExternal`.
1. Run defaulting functions on `SharedExternal` and `InstanceExternal`. **Implementations must take care to keep the defaults the same for fields that exist in both types. We should have tests for this.**
1. Convert `SharedExternal` to `SharedInternal` and `InstanceExternal` to `InstanceInternal`.
1. Decode flags over `SharedInternal` and `InstanceInternal` to enforce flag precedence. **Flags for fields that exist in both types must be registered against `InstanceInternal`, so that we maintain flag precedence when we merge the two. We should have a test for this too.**
1. Copy fields that exist in both types from `InstanceInternal` to `SharedInternal` so that the values from `InstanceInternal` are present everywhere. This is the merge.

Reply:
We need to sketch this out as integrated with the current flow, where flag precedence is applied and feature gates are set at multiple stages. I also didn't include validation in the above steps; in a simple scenario it just happens after 7, but validation also happens inside the dynamic config controller, so we likely need to inject some additional stuff there (similar to how we inject flag precedence today).